[Site Isolation] Remove child frames of old main frame when committing a provisional page

achristensen07 opened the pull request:
#### 7e5e04b40d3eae60d9cd0176061cf2a5ba92511d
<pre>
[Site Isolation] Remove child frames of old main frame when committing a provisional page
<a href="https://bugs.webkit.org/show_bug.cgi?id=274941">https://bugs.webkit.org/show_bug.cgi?id=274941</a>
<a href="https://rdar.apple.com/125374765">rdar://125374765</a>

Reviewed by Charlie Wolfe.

When a main frame navigates to a new domain, we need the iframes of the previous page to be
removed in the UI process and in all other processes.  This isn&apos;t an issue with site isolation
off because we cache or discard the entirety of the previous process when process swapping.
With site isolation, though, the same processes can be reused for the same domains after
main frame navigations.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::removeRemoteFrameProcessUse):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::commitProvisionalPage):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateToDomainAgainWithSPIUse)):

Canonical link: <a href="https://commits.webkit.org/279561@main">https://commits.webkit.org/279561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8ced7168ddd316c2de29cffeb93ba7582cceb16

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33192 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57102 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4547 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56128 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40686 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4406 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43581 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2977 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55922 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46553 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24720 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3863 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2702 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4054 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58697 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28992 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50994 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46699 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7964 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/29965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -536,6 +536,13 @@ void WebFrameProxy::setProcess(FrameProcess& process)
 
+void WebFrameProxy::removeRemoteFrameProcessUse(WebFrameProxy& mainFrame)
+{
+    m_frameProcess = mainFrame.m_frameProcess;
```
Why are you setting child frame processes to the main frame process? If we want to terminate child frame processes should we remove them from the browsing context group?

achristensen07 replied:
That wasn't great.  We need to remove the child frames in the UI process which will remove the references to the FrameProcesses which will then clean up.

charliewolfe commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -536,6 +536,11 @@ void WebFrameProxy::setProcess(FrameProcess& process)
 
+void WebFrameProxy::removeChildFrames()
+{
+    m_childFrames.clear();
```
Maybe this is fine for now, but shouldn't we keep the same-origin child frames?

achristensen07 replied:
Once back/forward cache is implemented we'll need to do something with them, but not keep them attached to the drawing area.

achristensen07 committed:
[Site Isolation] Remove child frames of old main frame when committing a provisional page
https://bugs.webkit.org/show_bug.cgi?id=274941
rdar://125374765

Reviewed by Charlie Wolfe.

When a main frame navigates to a new domain, we need the iframes of the previous page to be
removed in the UI process and in all other processes.  This isn't an issue with site isolation
off because we cache or discard the entirety of the previous process when process swapping.
With site isolation, though, the same processes can be reused for the same domains after
main frame navigations.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::removeRemoteFrameProcessUse):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::commitProvisionalPage):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateToDomainAgainWithSPIUse)):

Canonical link: https://commits.webkit.org/279561@main
